invisible_cities/detsim/simulate_S1.py

```diff
@@ -0,0 +1,188 @@
+        s1_pes_at_pmts = compute_S1_pes_at_pmts(x, y, z, s1_photons, S1_LT)
+        s1times = generate_S1_times_from_pes(s1_pes_at_pmts, time)
+        s1_wfs  = histogram_S1_times(s1times, buffer_length, wf_pmt_bin_width, tmin)
+        return s1_wfs
+        
+    return create_S1_waveforms_from_hits
+
+
+def histogram_S1_times(S1_times      : list,
+                       buffer_length : float,
+                       bin_width     : float,
+                       bias_time     : float=0)->np.ndarray:
+    """
+    S1_times are histogramed in a waveform of given buffer_length and bin_width
+    Parameters
+        :S1_times: list
+            output of generate_S1_times_from_pes, list of size equal to the
+            number of pmts. Each element is an array with the times of the S1
+    Returns:
+        :wfs: np.ndarray
+            waveforms with buffer_length and bin_width
+    """
+    bins = np.arange(0, buffer_length + bin_width, bin_width)
+    wfs  = np.stack([np.histogram(times-bias_time, bins=bins)[0] for times in S1_times])
+    return wfs
+
+
+def create_lighttable_function(filename : str)->Callable:
+    """ From a lighttable file, it returns a function of (x, y) for S2 signal
+    or (x, y, z) for S1 signal type. Signal type is read from the table.
+    Parameters:
+        :filename: str
+            name of the lighttable file
+    Returns:
+        :get_lt_values: Callable
+            this is a function which access the desired value inside
+            the lighttable. The lighttable values would be the nearest
+            points to the input positions. If the input positions are
+            outside the lighttable boundaries, zero is returned.
+            Input values must be vectors of same lenght, I. The output
+            shape will be (I, number_of_pmts).
+    """
+    lt     = load_dst(filename, "LT", "LightTable")
+    Config = load_dst(filename, "LT", "Config")    .set_index("parameter")
+    sensor = Config.loc["sensor"]     .value
+    lt     = lt.drop(sensor + "_total", axis=1) # drop total column
+
+    def get_lt_values(xs, ys, zs):
+        if len(zs) == 1:
+            zs = np.full(len(xs), zs)
+        if not (len(xs) == len(ys) == len(zs)):
+            raise Exception("input arrays must be of same shape")
+        xindices = pd.cut(xs, xbins, include_lowest=True)
+        yindices = pd.cut(ys, ybins, include_lowest=True)
+        zindices = pd.cut(zs, zbins, include_lowest=True)
+        indices = pd.Index(zip(xindices, yindices, zindices), name=("x", "y", "z"))
+        mask    = indices.isin(lt.index)
+        values  = np.zeros((len(xs), nsensors))
+        values[mask] = lt.loc[indices[mask]]
+        return values
+
+    if lt.get("z") is None:
+        lt["z"] = 1 # add fake z
+        had_z = False
+    else: had_z = True
+
+    lt = lt.set_index(["x", "y", "z"])
+    nsensors = lt.shape[-1]
+
+    xbins=binedges_from_bincenters(np.unique(lt.index.get_level_values('x')))
```

Comment:
There seems to be a numpy function that does this for you 
`histogram_bin_edges`, I believe you should call it with number of bins to be unique values +1, and specify the range you need.
However we need to decide what to do in case light table starts at `x_min> -r_active + x_pitch`, I think we should extend the lower and upper bin edge to be exactly on `r_active`. In that case we do need a custom function and `binedges_from_bincenters` can be easily adapted for this (ie adding arguments `left_edge, right_edge`). Eventually this function should then be moved to `core_functions.py`